daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/TailCallTest.scala

```diff
@@ -0,0 +1,150 @@
+           case (EQUAL @Int64 x 0) of
+               True -> 0
+             | _    -> ADD_INT64 x (F:triangle (SUB_INT64 x 1));
+
+         // Tail-recursive definition, via helper function
+         val triangleTR : (Int64 -> Int64) = F:triangleTR_acc 0;
+
+         // Tail-recursive definition using accumulator parameter
+         val triangleTR_acc : (Int64 -> Int64 -> Int64) = \ (acc: Int64) (x: Int64) ->
+           case (EQUAL @Int64 x 0) of
+               True -> acc
+             | _    -> F:triangleTR_acc (ADD_INT64 acc x) (SUB_INT64 x 1);
+
+
+         val triangle_viaFoldLeft : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDL @Int64 @Int64 ADD_INT64 0 (F:generate Nil@Int64 x);
+
+         val triangle_viaFoldRight : (Int64 -> Int64) = \ (x: Int64) ->
+            FOLDR @Int64 @Int64 ADD_INT64 0 (F:generate Nil@Int64 x);
+
+         // tail-recursive generator
+         val generate : (List Int64 -> Int64 -> List Int64) = \ (acc: List Int64) (x: Int64) ->
+           case (EQUAL @Int64 x 0) of
+               True -> acc
+             | _    -> F:generate (Cons @Int64 [x] acc) (SUB_INT64 x 1);
+
+       }
+      """
+
+  val small: Option[Int] = Some(5)
+  val unbounded: Option[Int] = None
+
+  "A *non* tail-recursive definition requires an unbounded env-stack, and an unbounded kont-stack" in {
+    val exp = e"F:triangle 100"
+    val expected = SValue.SInt64(5050)
+    // The point of this test is to prove that the bounded-evaluation checking really works.
+    runExpr(exp, envBound = unbounded, kontBound = unbounded) shouldBe expected
+    expectThrows {
+      runExpr(exp, envBound = small, kontBound = unbounded)
+    }
+    expectThrows {
+      runExpr(exp, envBound = unbounded, kontBound = small)
+    }
+  }
+
+  "A tail-recursive definition executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangleTR 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  "fold-left executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangle_viaFoldLeft 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  "fold-right executes with a small env-stack, and a small kont-stack" in {
+    val exp = e"F:triangle_viaFoldRight 100"
+    val expected = SValue.SInt64(5050)
+    runExpr(exp, envBound = small, kontBound = small) shouldBe expected
+  }
+
+  private def typeAndCompile(pkg: Ast.Package): PureCompiledPackages = {
+    val rawPkgs = Map(defaultParserParameters.defaultPackageId -> pkg)
+    Validation.checkPackage(rawPkgs, defaultParserParameters.defaultPackageId, pkg)
+    data.assertRight(
+      PureCompiledPackages(rawPkgs, Compiler.Config.Default.copy(stacktracing = FullStackTrace)))
+  }
+
+  val pkgs = typeAndCompile(pkg)
+
+  private def expectThrows[A](f: => A): Assertion = {
```

Comment:
This becomes probably obsolete if we use the `the[RuntimeException] thrownBy ...` approach outline above.